src/main/java/net/neoforged/neoforge/resource/ContextAwareReloadListener.java

```diff
@@ -5,15 +5,65 @@
 
 package net.neoforged.neoforge.resource;
 
+import com.google.gson.JsonElement;
+import com.mojang.serialization.JsonOps;
 import net.minecraft.core.RegistryAccess;
+import net.minecraft.resources.RegistryOps;
+import net.minecraft.server.packs.resources.PreparableReloadListener;
+import net.minecraft.server.packs.resources.SimplePreparableReloadListener;
+import net.neoforged.neoforge.common.conditions.ConditionalOps;
 import net.neoforged.neoforge.common.conditions.ICondition;
+import net.neoforged.neoforge.common.conditions.ICondition.IContext;
+import org.jetbrains.annotations.ApiStatus;
 
-public abstract class ContextAwareReloadListener {
+/**
+ * Reload listeners that descend from this class will have the reload context automatically populated when it is available.
+ * <p>
+ * The context is guaranteed to be available for the duration of {@link PreparableReloadListener#reload}.
+ * <p>
+ * For children of {@link SimplePreparableReloadListeners}, it will be available during both {@link SimplePreparableReloadListener#prepare prepare()} and {@link SimplePreparableReloadListener#apply apply()}.
+ */
+public abstract class ContextAwareReloadListener implements PreparableReloadListener {
+    /**
+     * TODO 1.20.5: Make private
+     * 
+     * @deprecated Use {@link #getContext()}
+     */
+    @Deprecated
     protected ICondition.IContext conditionContext = ICondition.IContext.EMPTY;
+
+    /**
+     * TODO 1.20.5: Make private
+     * 
+     * @deprecated Use {@link #getRegistryAccess()}
+     */
+    @Deprecated
     protected RegistryAccess registryAccess = RegistryAccess.EMPTY;
 
-    public void injectContext(ICondition.IContext conditionContext, RegistryAccess registryAccess) {
-        this.conditionContext = conditionContext;
-        this.registryAccess = registryAccess;
+    @ApiStatus.Internal
+    public void injectContext(ICondition.IContext context, RegistryAccess regAccess) {
+        this.conditionContext = context;
+        this.registryAccess = regAccess;
+    }
+
+    /**
+     * @return The context object held by this listener, or {@link IContext#EMPTY} if it is unavailable.
```

Comment:
Only `@return` is bad practice because the function won't have any javadoc in the summary. It's better to only say `Returns ...` than to only have `@return`.

Reply:
Yeah will need to move those to inline returns